### ISSUE TITLE
Investigate ways to better display callstack in Kibana

### ISSUE THREAD

**Issue by ryanrupp** [stagemonitor/stagemonitor#150]
A while back I was looking into ways to make the callstack easier to consume in Kibana similar to how the widget will display the callstack. Looking into this there was a feature added to Kibana to allow custom field formatters - https://www.elastic.co/blog/kibana-custom-field-formatters - as a proof of concept I basically added a "Stagemonitor Callstack" formatter which then takes the callstack JSON and turns it into a tree with D3 that can toggle the nodes open/close (see screeenshot at the bottom). At the time though I don't think the plugin API of Kibana was really developed out so it meant building from source, I'm not sure if maybe that's in a better state now but generally the idea of this issue would be to discuss ways in which the callstack when stored in Elasticsearch could be viewed in a rich way (toggle nodes, tooltips etc.) - my first thought was doing this in Kibana but alternatively could be done by any other web app that reads back the callstack JSON and displays it e.g. if the web widget had a mode where it could read requests back from Elasticsearch as the datasource. Currently I copy/paste the pretty print callstack into a text editor and look at it but with large SQL statements this becomes hard to follow. Also with this the pretty print stack would not need to be sent anymore possibly and just the JSON representation is stored in Elasticsearch.

<img width="1324" alt="kibanacustomcallstackstagemonitor" src="https://cloud.githubusercontent.com/assets/3022260/14889259/b70f765a-0d23-11e6-8d07-47c84ebedd95.png">

**Comment by felixbarny** [comment]
It's a really nice idea to use the field formatters for that! The plugin API of Kibana quite good now as far as I know. You can even package the extensions as a separate plug-in.

I have also thought about connecting the in browser widget to elasticsearch. There would probably be some issues with CORS though so we'd have to implement some sort of proxy servlet in stagemonitor. The most simple way of integration would be to just copy the request trace id into the widget. The most sophisticated integration would be a filterable and pageable table. But this would kind of be a duplication of what is available through the Kibana interface.

So I think the best would be to have an (optional) Stagemonitor Kibana plugin-in mid term and maybe implement a get-by-id mechanism in the widget for short term.

What do you think?

**Comment by ryanrupp** [comment]
Yeah I think an optional Kibana plugin would be the best for now (assuming there's ways to bundle this as a plugin now so it can be applied without having to build from source). For the widget approach I'd envision another servlet/endpoint being made available in Stagemonitor that forwards requests to Elasticsearch like you mentioned, this sounds more complicated/involved than a Kibana plugin though. Also in production we don't use the widget, although we could do the admin only enabling where you pass the admin password header to enable the widget.

I like the idea of the Kibana plugin for the callstack JSON and then not sending the pretty print format, which could be done through those hooks that were added recently to transform what is sent to Elasticsearch.

**Comment by felixbarny** [comment]
> Elasticsearch like you mentioned, this sounds more complicated/involved than a Kibana plugin though

The "get request trace by id" mechanism would actually be quite simple. Also you don't even need to enable the widget in production as you can just use a local or test server to paste the request trace id of the call tree you want to investigate.

**Comment by ryanrupp** [comment]
Looks like this can be achieved now possibly via the new plugin -
 https://github.com/stagemonitor/stagemonitor-kibana

**Comment by felixbarny** [comment]
Yes, indeed. But in the current version, only the ascii art tree is shown. We plan to use the same code as in the in browser widget to display the call tree.
